File: data-structures/queues/visual-basic/queue_circular.vb
' Raspberry Pi Foundation
' Usage licensed under CC BY-SA 4
' Note: This file is designed to be copied out and compiled on your machine.
' To run this file you need to:
' 1. Copy the contents
' 2. Paste them into the Visual Basic IDE of your choice (Visual Studio, for example)
' 3. Compile the program
' 4. Run the program


Module Program

    Public Const MAXSIZE As Integer = 4


    ' The Main subroutine is the default entry point for a VB program
    Sub Main()
        Console.WriteLine("### Circular queue ###")

        ' Initialise the queue and pointers
        Dim queue = New String(3) {}
        Dim front = -1
        Dim rear = -1

        ' Tuples are used to store multiple return values
        Dim enqueuedValues As Tuple(Of Integer, Integer) ' front, rear
        Dim dequeuedValues As Tuple(Of String, Integer, Integer) ' item, front, rear

        ' Insert test data into the queue
        enqueuedValues = Enqueue(queue, front, rear, "Julie")
        front = enqueuedValues.Item1
        rear = enqueuedValues.Item2
        OutputQueue(queue, front, rear)

        enqueuedValues = Enqueue(queue, front, rear, "Rey")
        front = enqueuedValues.Item1
        rear = enqueuedValues.Item2
        OutputQueue(queue, front, rear)

        enqueuedValues = Enqueue(queue, front, rear, "Habib")
        front = enqueuedValues.Item1
        rear = enqueuedValues.Item2
        OutputQueue(queue, front, rear)

        enqueuedValues = Enqueue(queue, front, rear, "Sabrina")
        front = enqueuedValues.Item1
        rear = enqueuedValues.Item2
        OutputQueue(queue, front, rear)

        ' Trying to enqueue - should say queue is full
        enqueuedValues = Enqueue(queue, front, rear, "Mina")
        front = enqueuedValues.Item1
        rear = enqueuedValues.Item2
        OutputQueue(queue, front, rear)

        ' Dequeue some of the items
        dequeuedValues = Dequeue(queue, front, rear)
        front = dequeuedValues.Item2
        rear = dequeuedValues.Item3
        Console.WriteLine($"Dequeued {dequeuedValues.Item1}")
        OutputQueue(queue, front, rear)

        dequeuedValues = Dequeue(queue, front, rear)
        front = dequeuedValues.Item2
        rear = dequeuedValues.Item3
        Console.WriteLine($"Dequeued {dequeuedValues.Item1}")
        OutputQueue(queue, front, rear)

        ' Insert data into the empty slots in the queue
        enqueuedValues = Enqueue(queue, front, rear, "Eirini")
        front = enqueuedValues.Item1
        rear = enqueuedValues.Item2
        OutputQueue(queue, front, rear)

        enqueuedValues = Enqueue(queue, front, rear, "Diane")
        front = enqueuedValues.Item1
        rear = enqueuedValues.Item2
        OutputQueue(queue, front, rear)

        ' Trying to enqueue - should say queue is full
        enqueuedValues = Enqueue(queue, front, rear, "Fergus")
        front = enqueuedValues.Item1
        rear = enqueuedValues.Item2
        OutputQueue(queue, front, rear)
    End Sub


    ' Check if the queue is empty
    Function IsEmpty(ByVal front As Integer) As Boolean
        If front = -1 Then
            Return True
        Else
            Return False
        End If
    End Function


    ' Check if the queue is full
    Function IsFull(ByVal front As Integer, ByVal rear As Integer) As Boolean
        If (rear + 1) Mod MAXSIZE = front Then
            Return True
        Else
            Return False
        End If
    End Function


    ' Enqueue an item
    Function Enqueue(ByVal queue As String(), ByVal front As Integer, ByVal rear As Integer, ByVal data As String) As Tuple(Of Integer, Integer)
        If IsFull(front, rear) = True Then
            Console.WriteLine($"Queue is full - {data} not added")
        Else
            rear = (rear + 1) Mod MAXSIZE
            queue(rear) = data
            If front = -1 Then front = 0 ' First item to be queued
        End If
        Return Tuple.Create(front, rear)
    End Function


    ' Dequeue an item
    Function Dequeue(ByVal queue As String(), ByVal front As Integer, ByVal rear As Integer) As Tuple(Of String, Integer, Integer)
        Dim dequeuedItem As String

        If IsEmpty(front) = True Then
            Console.WriteLine(vbLf & "Queue is empty - nothing to dequeue")
            dequeuedItem = ""
        Else
            dequeuedItem = queue(front)
            queue(front) = "" ' Testing
            ' Check if the queue is empty
            If front = rear Then
                front = -1
                rear = -1
            Else
                front = (front + 1) Mod MAXSIZE
            End If
        End If

        Return Tuple.Create(dequeuedItem, front, rear)
    End Function


    ' Output the state of the queue
    Sub OutputQueue(ByVal queue As String(), ByVal front As Integer, ByVal rear As Integer)
        ' Testing
        Console.Write(vbLf & "Queue: ")
        Console.WriteLine("[{0}]", String.Join(", ", queue))
        Console.WriteLine($"Front: {front}")
        Console.WriteLine($"Rear: {rear}")
    End Sub


End Module

File: data-structures/queues/java/queue_circular.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science
Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

import java.util.Arrays;

// A class to represent the data and pointer of a single item. The class is
// used to store multiple return values with different data types
class Item {
    private String data;
    private int frontPointer;
    private int rearPointer;

    public Item(String data, int frontPointer, int rearPointer) {
        this.data = data;
        this.frontPointer = frontPointer;
        this.rearPointer = rearPointer;
    }

    public String getData() {
        return data;
    }

    public int getFrontPointer() {
        return frontPointer;
    }

    public int getRearPointer() {
        return rearPointer;
    }
}


class QueuesCircularExample {
    public static final int MAX_SIZE = 4;

    // The main method is the entry point for all Java programs
    public static void main(String[] args) {
        System.out.println("### Circular queue ###");

        // Initialise the queue and pointers
        String[] queue = new String[MAX_SIZE];
        int front = -1;
        int rear = -1;

        // Array is used to store multiple return values of the same data type
        int[] enqueuedValues; // front, rear

        // Item object is used to store multiple return values with different data types
        Item dequeuedValues; // data, front, rear

        // Insert test data into the queue
        enqueuedValues = enqueue(queue, front, rear, "Julie");
        front = enqueuedValues[0];
        rear = enqueuedValues[1];
        outputQueue(queue, front, rear);

        enqueuedValues = enqueue(queue, front, rear, "Rey");
        front = enqueuedValues[0];
        rear = enqueuedValues[1];
        outputQueue(queue, front, rear);

        enqueuedValues = enqueue(queue, front, rear, "Habib");
        front = enqueuedValues[0];
        rear = enqueuedValues[1];
        outputQueue(queue, front, rear);

        enqueuedValues = enqueue(queue, front, rear, "Sabrina");
        front = enqueuedValues[0];
        rear = enqueuedValues[1];
        outputQueue(queue, front, rear);

        // Trying to enqueue - should say queue is full
        enqueuedValues = enqueue(queue, front, rear, "Mina");
        front = enqueuedValues[0];
        rear = enqueuedValues[1];
        outputQueue(queue, front, rear);

        // Dequeue some of the items
        dequeuedValues = dequeue(queue, front, rear);
        front = dequeuedValues.getFrontPointer();
        rear = dequeuedValues.getRearPointer();
        System.out.println("\nDequeued " + dequeuedValues.getData());
        outputQueue(queue, front, rear);

        dequeuedValues = dequeue(queue, front, rear);
        front = dequeuedValues.getFrontPointer();
        rear = dequeuedValues.getRearPointer();
        System.out.println("\nDequeued " + dequeuedValues.getData());
        outputQueue(queue, front, rear);

        // Insert data into the empty slots in the queue
        enqueuedValues = enqueue(queue, front, rear, "Eirini");
        front = enqueuedValues[0];
        rear = enqueuedValues[1];
        outputQueue(queue, front, rear);

        enqueuedValues = enqueue(queue, front, rear, "Diane");
        front = enqueuedValues[0];
        rear = enqueuedValues[1];
        outputQueue(queue, front, rear);

        // Trying to enqueue - should say queue is full
        enqueuedValues = enqueue(queue, front, rear, "Fergus");
        front = enqueuedValues[0];
        rear = enqueuedValues[1];
        outputQueue(queue, front, rear);
    }

    // Check if the queue is empty
    public static boolean isEmpty(int front) {
        if (front == -1)
            return true;
        else
            return false;
    }

    // Check if the queue is full
    public static boolean isFull(int front, int rear) {
        if ((rear + 1) % MAX_SIZE == front)
            return true;
        else
            return false;
    }

    // Enqueue an item
    public static int[] enqueue(String[] queue, int front, int rear, String data) {
        if (isFull(front, rear))
            System.out.println("\nQueue is full - " + data + " not added");
        else {
            rear = (rear + 1) % MAX_SIZE;
            queue[rear] = data;
            if (front == -1) // First item to be queued
                front = 0;
        }
        return new int[] {front, rear};
    }

    // Dequeue an item
    public static Item dequeue(String[] queue, int front, int rear) {
        String dequeuedItem;

        if (isEmpty(front)) {
            System.out.println("\nQueue is empty - nothing to dequeue");
            dequeuedItem = "";
        }
        else {
            dequeuedItem = queue[front];
            queue[front] = ""; // Testing
            // Check if the queue is empty
            if (front == rear) {
                front = -1;
                rear = -1;
            }
            else
                front = (front + 1) % MAX_SIZE;
        }

        return new Item(dequeuedItem, front, rear);
    }

    // Output the state of the queue
    public static void outputQueue(String[] queue, int front, int rear) {
        // Testing
        System.out.print("\nQueue: ");
        System.out.println(Arrays.toString(queue));
        System.out.println("Front: " + front);
        System.out.println("Rear: " + rear);
    }


}